Unify hybrid model tendencies and add rhoe_int

dennisYatunin opened the pull request:
This PR moves the explicit tendencies defined in `baroclinic_wave_utilities.jl` into `staggered_nonhydrostatic_model.jl`, unifying them into a single function and generalizing them to work with both 2D and 3D domains. It also separates the default tendencies from the non-default tendencies (e.g., hyperdiffusion and Held-Suarez forcing), so that test case files only need to specify non-default tendencies (if there are any), which can be done by defining methods for `additional_cache()` and `additional_tendency!()`.

In addition, this PR introduces a new naming convention for Fields defined on hybrid domains---Fields on cell centers are prefixed with a superscript "c", and Fields on cell faces are prefixed with a superscript "f". Previously, this was only done for some fields, and with non-superscript "c" and "f", which caused confusion when such fields were used alongside variables whose names start with "c" and "f" (e.g., heat capacity and Coriolis frequency). Similarly, this PR introduces a new data structure convention---Fields on cell centers are stored in `Y.c`, and Fields on cell faces are stored in `Y.f`. Previously, the prognostic state was stored as `Y.Yc.ρ`, `Y.Yc.ρe` (or equivalent), `Y.uₕ`, and `Y.w`, which was arbitrary and inefficient (for example, dss needed to be called on both `Y.Yc` and `Y.uₕ`, whereas it can now be called just once on `Y.c`). These new conventions account for most of the lines changed in this PR, but hopefully they will make the code easier to understand.

This PR also adds support for using internal energy as an prognostic variable, and it adds a new Held-Suarez test case that demonstrates this capability ("held_suarez_rhoe_int.jl"). It also adds the "inertial_gravity_wave.jl" test case to "hybrid/plane"; the analytic solution in this test case is still not working, so it can't be used as a proper integration test, but that can be fixed later.

Finally, this PR fixes a minor buildkite bug I introduced in my previous PR that was causing some artifacts from the rising bubble tests to be missing.

dennisYatunin commented:
bors try

jiahe23 commented on examples/hybrid/box/bubble_3d_invariant.jl:
```diff
@@ -321,7 +321,7 @@ function linkfig(figpath, alt = "")
 end
 
-dir = "bubble3d_invariant"
+dir = "bubble_3d_invariant"
```
```suggestion
dir = "bubble_3d_invariant_potential_temperature"
```

jiahe23 replied:
can you also change the name of the file adding 'potential_temperature' to be consistent with the total energy version.

jiahe23 commented on examples/hybrid/hyperdiffusion.jl:
```diff
@@ -0,0 +1,78 @@
+    ),
+    use_tempest_mode ? (; ᶠχw_data = similar(ᶠlocal_geometry, FT)) : (;),
+)
+
```
The unicode is over-used here. `div`, `grad`, `curl`, and their weak forms by adding a `w` in front is explanatory enough and very easy to read. I think we should be careful and not abuse unicode

jiahe23 commented on examples/hybrid/plane/bubble_2d_invariant.jl:
```diff
@@ -265,7 +265,7 @@ ENV["GKSwstype"] = "nul"
 Plots.GRBackend()
 
-dir = "bubble_invariant"
+dir = "bubble_2d_invariant"
```
```suggestion
dir = "bubble_2d_invariant_potential_temperature"
```

jiahe23 replied:
and the file name

jiahe23 commented on examples/hybrid/plane/inertial_gravity_wave.jl:
```diff
@@ -0,0 +1,364 @@
+    p₀(z) = p_0 * exp(-δ * z)
+end
+Tb_init(x, z) = ΔT * exp(-(x - xmid)^2 / d^2) * sin(π * z / zmax)
+T′_init(x, z) = Tb_init(x, z) * exp(δ * z / 2)
```
I am opposed to use `'` in var name

jiahe23 commented on examples/hybrid/plane/inertial_gravity_wave.jl:
```diff
@@ -0,0 +1,364 @@
+        kz = 2 * π / (2 * zmax) * ikz
+
+        # Fourier coefficient of ᶜρb_init (for current kx and kz)
+        ρ̂b_init = ρ̂b_init_array[ikx + max_ikx + 1, ikz + max_ikz + 1]
```
Please do not use `ρ̂` such notation

jiahe23 commented on examples/hybrid/hyperdiffusion.jl:
```diff
@@ -0,0 +1,78 @@
+    (; ᶜp, ᶜχ, ᶜχuₕ, κ₄, use_tempest_mode) = p # assume that ᶜp has been updated
+    point_type = eltype(Fields.local_geometry_field(axes(Y.c)).coordinates)
+
+    if use_tempest_mode
```
tempest mode does contain the total energy and internal energy set up.

jiahe23 commented on examples/hybrid/hyperdiffusion.jl:
```diff
@@ -0,0 +1,78 @@
+            @. ᶜχ = W∇◦ₕ(∇ₕ(Y.c.ρθ)) # ᶜχρθ
+            Spaces.weighted_dss!(ᶜχ)
+            @. Yₜ.c.ρθ -= κ₄ * W∇◦ₕ(∇ₕ(ᶜχ))
+        elseif :ρe in propertynames(Y.c)
+            @. ᶜχ = W∇◦ₕ(∇ₕ(Y.c.ρe + ᶜp)) # ᶜχρe
+            Spaces.weighted_dss!(ᶜχ)
+            @. Yₜ.c.ρe -= κ₄ * W∇◦ₕ(∇ₕ(ᶜχ))
+        elseif :ρe_int in propertynames(Y.c)
+            @. ᶜχ = W∇◦ₕ(∇ₕ(Y.c.ρe_int + ᶜp)) # ᶜχρe_int
+            Spaces.weighted_dss!(ᶜχ)
+            @. Yₜ.c.ρe_int -= κ₄ * W∇◦ₕ(∇ₕ(ᶜχ))
```
no need for these two and they do not seem correct.

jiahe23 commented on examples/Project.toml:
```diff
@@ -18,12 +15,16 @@ LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 Logging = "56ddb016-857b-54e1-b83d-db4d58db5568"
 OrdinaryDiffEq = "1dea7af3-3e70-54e6-95c3-0bf5283fa5ed"
 Plots = "91a5bcdd-55d7-5caf-9e0b-520d859cae80"
+ProgressLogging = "33c8b6b6-d38a-422a-b730-caa89a2f386c"
```
where do we need this?

dennisYatunin replied:
This is used for progress bars in `inertial_gravity_wave.jl`. Generating the analytical solution takes a very long time, so I added `@progress` to some for loops.

jiahe23 reviewed (changes requested):
Overall looks good. I like the idea of make it clear between center and face. 

Here are some issues:

hyperdiffusion file needs to revise a bit, @szy21 can you also take a close look at this file just in case I missed something.

unicode is over-used. Here are two examples:
- `ρ` is good, but `ρ̂ ` is too much; and we used to have reports on  superscripts/subscripts/overbars not appearing properly in some browser;
- `hdiv` and `hwdiv` is self-explanatory enough and the symbols only make it hard to read.

szy21 commented:
I agree with Jia that unicode is overused. I’m fine with doing this in climacore for now but let’s not do it in climaatmos. It would be good to talk about guidelines on unicodes before we make changes in climaatmos.

jiahe23 commented:
> I agree with Jia that unicode is overused. I’m fine with doing this in climacore for now but let’s not do it in climaatmos. It would be good to talk about guidelines on unicodes before we make changes in climaatmos.

Yep, let's make minimal changes here and have it in for implicit development; but bear in mind the unicode issue.

dennisYatunin commented:
I've addressed the above comments as follows:
1. I renamed all the rising bubble filenames to follow the same convention as the sphere filenames, so that they now end with `_rhoe` or `_rhotheta`. I also updated the output directories and buildkite pipeline appropriately.
2. I made the Unicode replacements `∇◦ → div`, `W∇◦ → wdiv`, `∇⨉ → curl`, `W∇⨉ → wcurl`, `∇ → grad`, `W∇ → wgrad`, and `ρ̂ → ρf` (and likewise for all the other variables with hats). Also, although this was not requested, I made the replacement `I → interp`, since `I` almost always means "identity". I did not replace the `′` symbol, since it is already in use in ClimaCore's source code and is a very common mathematical symbol. Please let me know if you feel that is a bad decision; I can replace `′` with `prime`, but I don't feel that this is necessary.
3. I replaced the ρe and ρe_int cases in tempest-mode hyperdiffusion with `error("use_tempest_mode must be false when not using ρθ")`.

In general, how does the following work as a Unicode rule of thumb:
"Only common mathematical symbols, non-English alphabets, English/numerical superscripts, and English/numerical subscripts may be used."

dennisYatunin commented:
bors try

jiahe23 commented:
> I've addressed the above comments as follows:
> 
> 1. I renamed all the rising bubble filenames to follow the same convention as the sphere filenames, so that they now end with `_rhoe` or `_rhotheta`. I also updated the output directories and buildkite pipeline appropriately.
> 2. I made the Unicode replacements `∇◦ → div`, `W∇◦ → wdiv`, `∇⨉ → curl`, `W∇⨉ → wcurl`, `∇ → grad`, `W∇ → wgrad`, and `ρ̂ → ρf` (and likewise for all the other variables with hats). Also, although this was not requested, I made the replacement `I → interp`, since `I` almost always means "identity". I did not replace the `′` symbol, since it is already in use in ClimaCore's source code and is a very common mathematical symbol. Please let me know if you feel that is a bad decision; I can replace `′` with `prime`, but I don't feel that this is necessary.
> 3. I replaced the ρe and ρe_int cases in tempest-mode hyperdiffusion with `error("use_tempest_mode must be false when not using ρθ")`.
> 
> In general, how does the following work as a Unicode rule of thumb: "Only common mathematical symbols, non-English alphabets, English/numerical superscripts, and English/numerical subscripts may be used."

Thanks! It looks good.

We could leave the prime for later decision. I find it confusing with its common use in matrix transpose.

dennisYatunin commented:
bors r+

dennisYatunin commented:
bors r+
